### ISSUE TITLE
[bug] Unable to spin up 200+ dask workers on AWS QHub

### ISSUE THREAD

**Issue by iameskild** [Quansight/qhub#828]
## Describe the bug
A clear and concise description of what the problem is.

When requesting a large number of dask workers, some fraction of worker pods never spin up. During some reason testing, I requested 250 workers and was only ever to get 190 (or so) workers to spin up; 190 (or thereabouts) seems to be the limit of dask workers nodes we can spin up at one time. For the node-group in question, we have set the max number of workers to the default 450 in the `qhub-config.yaml`.  

Checking the status of these nodes on AWS reveals an interesting error message:
```
runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

## How can we help?
Help us help you.
* What are you trying to achieve?
* How can we reproduce the problem?
* What is the expected behaviour?
* And what is currently happening?
* Any error messages?
> If helpful, please add any screenshots and relevant links to the description.

Here are a few screenshots of these error messages:

![Screen Shot 2021-09-20 at 16 30 50](https://user-images.githubusercontent.com/42120229/134090433-dac5cbd9-8e2a-4428-8248-3549c1976025.png)
<img width="1661" alt="Screen Shot 2021-09-20 at 16 31 55" src="https://user-images.githubusercontent.com/42120229/134090435-39991ce0-85c4-4e1a-a457-7441f314b9ba.png">

## Your environment
Describe the environment in which you are experiencing the bug.
> Include your conda version (use `conda --version`), k8s and any other relevant details.
```
qhub - 0.3.12 (installed from commit 0dff706)
k8s - 1.19
```

**Comment by Adam-D-Lewis** [comment]
AWS has limits on the maximum number of ec2 instances you can spin up.  You can increase these limits by request.  Have you checked if this is the issue?

**Comment by iameskild** [comment]
Hey @Adam-D-Lewis, thanks for following up so quickly. 

We have set the maximum number of nodes for this specific node-group to the 450 (in the `qhub-config.yml`), which I believe is the [default service quota](https://docs.aws.amazon.com/eks/latest/userguide/service-quotas.html). The problem is that we still are limited to some fraction of these 450 nodes. And from my screenshots above and from my experience, it looks like the EC2 instances spin up but the dask worker pods never do. 

I found [this EC2 services quota page](https://docs.aws.amazon.com/general/latest/gr/ec2-service.html) but it's not clear if there is a limit we are hitting.

**Comment by Adam-D-Lewis** [comment]
In the screenshot above, it looks like the nodes are in the "Not Ready" state.  Do they stay that way or do they eventually go to "Ready"?  It seems that if the nodes aren't becoming ready, the issue may be with the docker image (seems unlikely because other workers are spinning up fine), or possibly something on the AWS side.  

Doing some quick googling brought up this link https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html. 
 Could you check out the section titled "Container runtime network not ready", try the recommended fix and see if that works for you?

**Comment by Adam-D-Lewis** [comment]
This might be relevant as well https://aws.amazon.com/premiumsupport/knowledge-center/eks-cni-plugin-troubleshooting/

**Comment by iameskild** [comment]
The nodes remain in the "Not Ready" state and from my experience, never become "Ready". 
Thanks for sharing that link, I confirmed that the Role ARN is in the configuration map, not the Instance Profile ARN.

I'm taking a look at the second link now :)

**Comment by iameskild** [comment]
cc: @brl0 @Adam-D-Lewis 

It's taken me some time but I believe the underlying problem here is the limited number of IP addresses in each of the subnets. I requested 250 dask-workers (each of which is running on a separate node, in our case on a `r5.4xlarge` instance type), and after waiting 20 minutes for a quorum of 200, I seemed to be stuck with 174 workers. 

I decided to check the number of available IP address for each of the subnets and these were the results:
```
> aws ec2 describe-subnets | jq '.Subnets[] | .SubnetId + "=" + "\(.AvailableIpAddressCount)"'

"subnet-xxxd7b=0"
"subnet-xxxe5f=0"

```

A few months ago we updated our CIDR newbits to increase the number of available IPs to 4091 [`2**(32 - newbits) -5`] and I naively assumed that would be plenty. However reading more about how AWS allocates IPs through the use of ENIs, each of [our nodes can consume up to 60 IP addresses](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI): 
```
30 IPs * 2 ENIs = 60 IPs
```

From what I can tell, if any of the IPs are used in the first ENI, a whole new ENI is reserved for that node, hence 30 times 2. And I believe this is controlled through the `WARM_ENI_TARGET=1`. I have confirmed that this value is set to 1 for the node-group in question. 

There are other variables that might help bring down the number of IPs being used, namely using `WARM_IP_TARGET`. [Here is a brief discussion ](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/docs/eni-and-ip-target.md) on the tradeoffs between reducing the number of available IPs per node (namely possibly hitting AWS EC2 API limit) versus reducing the number of nodes that can be launched (my current issue).  

Given that we run one dask-worker pod on each node (so 3 pods per node), I don't foresee us hitting the AWS EC2 API limit, but I'm happy to hear from others. 

From [this medium article](https://betterprogramming.pub/amazon-eks-is-eating-my-ips-e18ea057e045), a quick work around might be to set the `WARM_IP_TARGET=2`:
```
kubectl -n kube-system set env daemonset aws-node WARM_IP_TARGET=2
```
I have not tested this approach yet and I'm curious to hear from others on what a long(er) solution might look like. 

Thanks again for any feedback and suggestions.

**Comment by costrouc** [comment]
@iameskild lets make these vpcs as large as we can then. I see for aws the vpc is "10.10.0.0/16" and thus `2^16 = 65536` ips and we reserve the first `4` bits for the subnets -> thus `2^12 = 4096`. Instead lets use the entire ip `10.0.0.0/8` and use `4` bits for the subnet.

**Comment by iameskild** [comment]
@costrouc That sounds like a good idea! I'll submit a PR with that change shortly

**Comment by Adam-D-Lewis** [comment]
I'm a complete network newbie, but I don't think this will work.  In AWS documentation it says the following:

"When you create a VPC, you must specify an IPv4 CIDR block for the VPC. The allowed block size is between a /16 netmask (65,536 IP addresses) and /28 netmask (16 IP addresses). After you've created your VPC, you can associate secondary CIDR blocks with the VPC. For more information, see Add IPv4 CIDR blocks to a VPC."

So it looks like you can specify secondary/16 CIDR blocks, but not a single /8 block.

**Comment by costrouc** [comment]
@Adam-D-Lewis yeah looks that you are right. Well what if we just reduce the number of subnet bits then to `2`. This would only allow 4 subnets and would also quadruple the number of nodes we could support so this would be around 700 nodes.

**Comment by dharhas** [comment]
So what about fixing this at the source? i.e. if we are trying to allocate 250 IPs, why are we requesting more than that many IPs? If we don't fix that then everything else is a bandaid as we try and use larger and larger clusters.
